I doubt Josh will approve of this but w/e

time-killer-games opened the pull request:
I'm tired of maintaining two different copies of this code, one for enigma, and one for the game maker extension. For now on, if anyone wants to have access to the latest features of dialog module, they will need to download my dll, install libffi, and enable external functions with this widget system selected from the dropdown. The other widget systems will still get bug fixes, but I want all my focus to be on updating the dll so both enigma and gamemaker users can use something equally up-to-date at any given time. I spend way too much time messing with dialogs, this will cut down the need for maintenance by at least a third. Because only the dll will be getting new bells and whistles whenever I choose to do that; not the other systems. If I had my way, all our widget systems would be deleted except this one. (Can I *please* do this?)

Josh thinks using a dll is a dumb idea, but guess what? ENIGMA uses DLL's already, a crap ton of them, or at least on linux and bsd. Windows and mac are statically linked, but they are still third-party dependencies nonetheless. I don't see how this is any different, to be honest...

This will also enable me to make new changes to the library without even needing to make a new enigma pull request - as in this way the changes i can make in my own repository.

GameMaker extensions that call DLL's without using scripts will fail silently if the dll is not found. I figured since YoYo is doing this, and since I need this to happen for these changes, I see no harm in it just being printed instead, after all, fundies said, "terminal printing is better than GUI".

```
git clone -b patch-54 https://github.com/time-killer-games/enigma-dev-fork.git ~/enigma-dev
git clone https://github.com/time-killer-games/DialogModule.git # for building from source.
```
Example GM81 Project + Required Library: [DialogModule.zip](https://github.com/enigma-dev/enigma-dev/files/4778500/DialogModule.zip)

time-killer-games commented:
@RobertBColton how would i go about making the compiler do this with an exported game?

For example, if the user is on Windows and building 64-bit with this Widget system enabled:
```
curl https://github.com/time-killer-games/DialogModule/raw/master/DialogModule.dll/DialogModule/DialogModule%20(x64)/DialogModule.dll --output "C:/path/to/game/directory/DialogModule.dll"
```

RobertBColton commented:
You can run plain old shell commands in the makefile, so what you already put there is probably fine. I understand you are missing the path to the output executable, you'll need to look at our existing `SHELL/Makefile` to find it. Hypothetically I see nothing stopping you from making this work.

time-killer-games commented:
Hey, that's pretty awesome. That's a lot less involved and much easier than I thought it would be. I though I would have to edit the source code for emake and the plugin, but this changes everything. Thank you so much for pointing me in the right direction! :D

RobertBColton commented:
Alright, I am going to do an actual assessment of this pull request to avoid wasting more of your time.

The reason Josh is unlikely to approve this is that it doesn't help meet several of ENIGMA's design goals. For example: ENIGMA should work without an internet connection once dependencies are downloaded; the ENIGMA project needs to be able to control where dependencies are downloaded from; ENIGMA users should be able to build standalone executables that use dialogs; etc.

I have no problem with you maintaining this approach yourself. It just doesn't belong in the main ENIGMA repository, as anything there needs to meet the above goals (something wget also violates). Therefore, I am going to close this pull request- feel free to open a new one if you would like to try a different approach.

I understand you would like to maintain only a single copy of your dialogmodule code. I think we may be able to come up with a solution to that problem that also meet's enigma's goals, if you are interested.

time-killer-games commented:
Well, it can easily be done on Mac, Linux, and BSD. But the Windows verion relies on MSVC features such as ATL for the interpreted VBScript InputBox. The reason I even use that is because that is the only option Microsoft provides for C++ projects that is an input box dialog which resizes to fit the contents of the text label above the textbox, to ensure it doesn't truncate or run out of room in the dialog's client area. MS doesn't provide any easy way to make a text label resize a dialog to ensure it fits, so this is a workaround to that problem, even yoyogames has struggled to implement this themselves on Windows. They use a custom dialog they wrote themselves which doesn't resize correctly when the text string is long enough in the static label control.

time-killer-games commented:
@RobertBColton so that being said, should I just wait until we fully support compiling with MSVC as my only real option here?

RobertBColton commented:
Yeah, though I don't think that sizing the dialog to fit its contents is necessarily as hard as you think. It's very easy to get the text extents and then factor that into the initial size we apply in the dialog procedure.
https://stackoverflow.com/a/11600004

time-killer-games commented:
@RobertBColton you must've not read a lot of that stackoverflow answer, It doesn't take new lines and a lot of other vital info into account, making the approach suggested utterly useless.

RobertBColton commented:
DrawText handles newlines just fine.
https://stackoverflow.com/a/34189620

time-killer-games commented:
@RobertBColton you aren't reading still, clearly, i wasn't even talking about that function

time-killer-games commented:
@RobertBColton fundies already talked about splitting up the main executable into multiple files, where there is an executable that is dependent on an external resources file. That much Josh approved of, so allow me to explain other reasons why what you said doesn't make sense:

I can easily change the installer script from the wiki to download the dll's with base enigma, then a simple cp command can place the dll into the working directory for the target executable. i used to have a widget system that was windows-only that was exactly like this pull request and josh merged it into master a while back. 

The reason the system was removed was because it was windows only and because our dll calling functions were broken at the time. The reason i said "i doubt josh would approve of this" is because he is a person who often changes his mind for what he is/isn't ok with for no apparent reason quite often, and he has a history of doing this that I've observed.

time-killer-games commented:
@JoshDreamland you say I have a memory problem, but the real problem is we prioritize things we value as important to remember differently. There was a time we had a widget system just like this, but we removed it because dll's were broken for a long ass time we eventually gave up on it ever being fixed which rendered the widget system useless so we removed it. I remember talking with you about adding this system back and making it cross-platform now that dlls work, and you didn't approve and didnt even know what I was talking about when I said "add it back". Do you not simply remember anything I just said that happened and that is your only reason for changing your mind?

If you don't want me constantly making pull requests to update widgets - this is meant to save all of us that same head ache - i could just maintain the shit in my own repository and you won't see another dialog pull request ever again. I think that is good thing for everyone involved.
